// axios/index.js
async function request(method, url, data) {
  const options = { method };
  if (data !== undefined) {
    options.headers = { 'Content-Type': 'application/json' };
    options.body = JSON.stringify(data);
  }
  const response = await fetch(url, options);
  const text = await response.text();
  let parsed;
  try {
    parsed = JSON.parse(text);
  } catch (err) {
    parsed = text;
  }
  return {
    data: parsed,
    status: response.status,
    statusText: response.statusText,
    headers: response.headers,
  };
}

module.exports = {
  get(url, config = {}) {
    return request('GET', url, undefined, config);
  },
  post(url, data, config = {}) {
    return request('POST', url, data, config);
  },
};

function buildClient(baseURL = '') {
  return {
    get(path, config = {}) {
      return request('GET', `${baseURL}${path}`, undefined, config);
    },
    post(path, data, config = {}) {
      return request('POST', `${baseURL}${path}`, data, config);
    },
  };
}

module.exports.create = function create(config = {}) {
  const baseURL = config.baseURL || '';
  return buildClient(baseURL);
};
